src/java/picard/analysis/replicates/CollectIndependentReplicateMetrics.java

```diff
@@ -0,0 +1,539 @@
+        }
+        if (useLocus && newLocus) {
+            metric.merge(locusData);
+            log.debug("Merged final metric. nSites:" + metric.nSites);
+        } else {
+            metric.nThreeAllelesSites++;
+            log.debug("didn't merge last metric, due to 3rd allele: nThreeAllelesSites =" + metric.nThreeAllelesSites);
+        }
+
+        log.info("Iteration done. Emitting metrics.");
+
+        // Emit metrics
+        final MetricsFile<IndependentReplicateMetric, Byte> metricsFile = getMetricsFile();
+
+        metric.calculateDerivedFields();
+        metricsFile.addMetric(metric);
+        metricsFile.addHistogram(alleleBalanceCount);
+        metricsFile.addHistogram(umiEditDistanceInDiffBiDups);
+        metricsFile.addHistogram(umiEditDistanceInSameBiDups);
+
+        metricsFile.write(OUTPUT);
+
+        final MetricsFile<?, ComparableTuple<String, String>> confusionMetrics = getMetricsFile();
+
+        confusionMetrics.addHistogram(umiConfusionMatrix);
+        confusionMetrics.addHistogram(umiConfusionMatrixEditDistance);
+        confusionMetrics.write(MATRIX_OUTPUT);
+
+        return 0;
+    }
+
+    private enum SetClassification {
+        MISMATCHING_ALLELE,
+        DIFFERENT_ALLELES,
+        REFERENCE_ALLELE,
+        ALTERNATE_ALLELE
+    }
+
+    /**
+     * a small utility to inform if one interval is cleanly before another, meaning that they do not overlap and
+     * the first is prior (in genomic order) to the second
+     *
+     * @param lhs the "first" {@link QueryInterval}
+     * @param rhs the "second" {@link QueryInterval}
+     * @return true if the to intervals do not intersect _and_ the first is prior to the second in genomic order
+     */
+    private static boolean isCleanlyBefore(final QueryInterval lhs, final QueryInterval rhs) {
+        return !lhs.overlaps(rhs) && lhs.compareTo(rhs) < 0;
+    }
+
+    private static SetClassification classifySet(final int nRef, final int nAlt, final int nOther) {
+        // if we found any "other" alleles, this is a mismatching set
+        if (nOther != 0) return SetClassification.MISMATCHING_ALLELE;
+
+        // if we found both ref and alt alleles, this is a heterogeneous set
+        if (nAlt > 0 && nRef > 0) return SetClassification.DIFFERENT_ALLELES;
+
+        //if we found no reference alleles, this is an "alternate" set
+        if (nRef == 0) return SetClassification.ALTERNATE_ALLELE;
+
+        //if we found no alternate alleles, this is a "reference" set.
+        if (nAlt == 0) return SetClassification.REFERENCE_ALLELE;
+
+        throw new IllegalAccessError("shouldn't be here!");
+    }
+
+    private static QueryInterval queryIntervalFromSamRecord(final SAMRecord samRecord) {
+        return new QueryInterval(samRecord.getReferenceIndex(), samRecord.getStart(), samRecord.getEnd());
+    }
+
+    /** Gives the edit distance between this barcode and another of the same length. */
```

Comment:
Is this ever enforced anywhere (that they are the same length)?  Do you think it's important to?